### PR TITLE
Improvements to Model Reduction

### PR DESCRIPTION
**Pull request by ngoiz**
Includes:
- Bug fixes and documentation for Balancing ROMS
- Single-sided Krylov Projection (using `single_side = observability` when building the Krylov ROM greatly improves performance!)
- Documentation improvements to linear modules

### CODE REVIEW

File: sharpy/postproc/asymptoticstability.py [hunk 1]
```diff
@@ -216,10 +216,14 @@ def print_eigenvalues(self):
 
     def velocity_analysis(self):
 
-        cout.cout_wrap('Velocity Asymptotic Stability Analysis')
-
         ulb, uub, num_u = self.settings['velocity_analysis']
 
+        if self.settings['print_info']:
+            cout.cout_wrap('Velocity Asymptotic Stability Analysis')
```
Comment (fonsocarre) [inline sharpy/postproc/asymptoticstability.py]:
```suggestion
            cout.cout_wrap('Velocity Asymptotic Stability Analysis', 1)
```

File: sharpy/postproc/asymptoticstability.py [hunk 2]
```diff
@@ -216,10 +216,14 @@ def print_eigenvalues(self):
...
+            cout.cout_wrap('Initial velocity: %.2f m/s' % ulb, 1)
+            cout.cout_wrap('Final velocity: %.2f m/s' % uub, 1)
+            cout.cout_wrap('Number of evaluations: %g' %num_u, 1)
```
Comment (fonsocarre) [inline sharpy/postproc/asymptoticstability.py]:
```suggestion
            cout.cout_wrap('Number of evaluations: %g' % num_u, 1)
```

File: sharpy/postproc/frequencyresponse.py [hunk 3]
```diff
@@ -137,8 +137,8 @@ def run(self):
             else:
                 compute_fom = True
 
-        cout.cout_wrap('Computing frequency response...')
         if (self.settings['compute_fom'].value and self.settings['load_fom'] == '') or compute_fom:
+            cout.cout_wrap('Computing frequency response...')
```
Comment (fonsocarre) [inline sharpy/postproc/frequencyresponse.py]: Shouldn't the `cout` stuff be wrapped in a `if self.settings['print_info']`?

File: sharpy/rom/balanced.py [hunk 4]
```diff
@@ -103,7 +124,10 @@ def run(self, ss):
                                     gv=S,
                                     convergence=self.settings['convergence'],
                                     method=self.settings['reduction_method'])
-
+            if librom.check_stability(ssrom.A, dt=True):
+                cout.cout_wrap('ROM by direct balancing is stable')
```
Comment (fonsocarre) [inline sharpy/rom/balanced.py]: These things should definitely be quiet (or wrapped in an `if self.settings['print_info']` for the good outcome (stable ROM). I understand if the output is forced when it is unstable.

File: sharpy/rom/utils/krylovutils.py [hunk 5]
```diff
@@ -428,3 +479,17 @@ def remove_a12(As, n_stable):
...
+
+
+def check_eye(T, Tinv, msg=''):
+    """Simple utility to verify matrix inverses"""
+    eye_approx = Tinv.dot(T)
+    max_diff = np.max(np.abs(np.eye(eye_approx.shape[0]) - eye_approx))
+
+    try:
+        log_error = np.log10(max_diff)
+        assert log_error < -6, 'T.dot(Tinv) not equal to identity, %s \nlog(error) = %.e' \
```
Comment (fonsocarre) [inline sharpy/rom/utils/krylovutils.py]: Set the threshold for convergence as an optional argument in the function. If you want to deal with log of the error, please use a variable name such as `log_error` to make explicit what you are talking about.

File: sharpy/rom/utils/librom.py [hunk 6]
```diff
@@ -1250,6 +1250,28 @@ def eigen_dec(A, B, C, dlti=True, N=None, eigs=None, UR=None, URinv=None,
...
+        dt (bool): Discrete time system
+
+    Returns:
+        bool: True if the system is stable
+    """
+    eigvals = scalg.eigvals(A)
+    if dt:
+        criteria = np.abs(eigvals) > 1.
+    else:
+        criteria = np.real(eigvals) > 0
```
Comment (fonsocarre) [inline sharpy/rom/utils/librom.py]:
```suggestion
        criteria = np.real(eigvals) > 0.0
```

File: sharpy/rom/utils/librom.py [hunk 7]
```diff
@@ -1250,6 +1250,28 @@ def eigen_dec(A, B, C, dlti=True, N=None, eigs=None, UR=None, URinv=None,
...
+
+    if np.sum(criteria) >= 1:
```
Comment (fonsocarre) [inline sharpy/rom/utils/librom.py]:
```suggestion
    if np.sum(criteria) >= 1.0:
```

File: sharpy/rom/utils/librom.py [hunk 8]
```diff
@@ -1250,6 +1250,28 @@ def eigen_dec(A, B, C, dlti=True, N=None, eigs=None, UR=None, URinv=None,
...
+        return True
+    else:
+        return False
+
 if __name__ == '__main__':
```
Comment (fonsocarre) [inline sharpy/rom/utils/librom.py]: Can we cleanup this?

File: sharpy/postproc/frequencyresponse.py (hunk 3 above)
Comment (ngoiz) [reply sharpy/postproc/frequencyresponse.py]: Indeed, preparing a new commit

File: sharpy/rom/utils/librom.py (hunk 8 above)
Comment (ngoiz) [reply sharpy/rom/utils/librom.py]: new test case added

File: sharpy/rom/balanced.py (hunk 4 above)
Comment (ngoiz) [reply sharpy/rom/balanced.py]: Included in the latest commits

**Pull request merged**